src/vm/index.md

```diff
@@ -94,38 +95,63 @@ Then the following registers are initialized (without explicit initialization, a
 
 - External: predicate and script. `$fp` will be zero.
 - Internal: call. `$fp` will be non-zero.
 
 [Returning](./instruction_set.md#return-return-from-call) from a context behaves differently depending on whether the context is external or internal.
 
+## Predicate Estimation
+
+For any input of type [`InputType.Coin`](../protocol/tx_format/index.md) or [`InputType.Message`](../protocol/tx_format/index.md), a non-zero `predicateLength` field means the UTXO being spent is a [P2SH](https://en.bitcoinwiki.org/wiki/P2SH) rather than a [P2PKH](https://en.bitcoinwiki.org/wiki/Pay-to-Pubkey_Hash) output.
+
+For each such input in the transaction, the VM is [initialized](#vm-initialization), then:
+
+1. `$pc` and `$is` are set to the start of the input's `predicate` field.
+1. `$ggas` and `$cgas` are set to `tx.gasLimit`.
+
+Predicate estimation will fail if gas is exhausted during execution.
+
+During predicate mode, hitting any of the following instructions causes predicate estimation to halt, returning Boolean `false`:
+
+1. Any [contract instruction](./instruction_set.md#contract-instructions).
+
+In addition, during predicate mode if `$pc` is set to a value greater than the end of predicate bytecode (this would allow bytecode outside the actual predicate), predicate estimation halts returning Boolean `false`.
+
+A predicate that halts without returning Boolean `true` would not pass verification, making the entire transaction invalid. Note that predicate validity is monotonic with respect to time (i.e. if a predicate evaluates to `true` then it will always evaluate to `true` in the future).
+
+After successful execution, `predicateGasUsed` is set to `tx.gasLimit - $ggas`.
+
 ## Predicate Verification
 
 For any input of type [`InputType.Coin`](../protocol/tx_format/input.md#inputcoin) or [`InputType.Message`](../protocol/tx_format/input.md#inputmessage), a non-zero `predicateLength` field means the UTXO being spent is a [P2SH](https://en.bitcoinwiki.org/wiki/P2SH) rather than a [P2PKH](https://en.bitcoinwiki.org/wiki/Pay-to-Pubkey_Hash) output.
 
 For each such input in the transaction, the VM is [initialized](#vm-initialization), then:
 
-1. `$pc`  and `$is` are set to the start of the input's `predicate` field.
+1. `$pc` and `$is` are set to the start of the input's `predicate` field.
+1. `$ggas` and `$cgas` are set to `predicateGasUsed`.
+
+Predicate verification will fail if gas is exhausted during execution.
 
 During predicate mode, hitting any of the following instructions causes predicate verification to halt, returning Boolean `false`:
 
 1. Any [contract instruction](./instruction_set.md#contract-instructions).
```

Comment:
Should this be reworded a bit since there is no list anymore - it's only one item. Maybe something like

> During predicate mode, hitting any [contract instruction](./instruction_set.md#contract-instructions) causes predicate verification to halt, returning Boolean `false`.

Reply:
I agree, I'll get this in now.

Reply:
done